onnxruntime/core/session/inference_session.cc

```diff
@@ -104,8 +104,11 @@ InferenceSession::InferenceSession(const SessionOptions& session_options, loggin
   // Beyond this, we will create a global thread pool to share across sessions.
   {
     int pool_size = session_options_.session_thread_pool_size <= 0
-                        ? std::thread::hardware_concurrency() / 2
+                        ? std::thread::hardware_concurrency()
                         : session_options_.session_thread_pool_size;
+    /// If we can't determine the number of cores, a pool size of 1 will maintain sequential execution
+    if (pool_size <= 0)
+      pool_size = 1;
```

Comment:
This thread doesn't really get used as when NumThreads() == 1, the fn is executed in the calling thread? Wondering if we should still create it? Are we creating it to avoid all call sites to check for nullptr?

Reply:
Yes.  A ParallelForAsync() will give up the current thread and we will still need the 1 thread in the pool to run the unit of work.  So we want it for consistency.  The short circuit in the code was added for a slight performance gain, not by design.

Reply:
Sure.